proposals/kroxylicious-operator-api-alternative.md

```diff
@@ -0,0 +1,652 @@
+
+The VirtualCluster CR is the responsibilty of the the Developer - As per current implementation.
+
+```yaml
+apiVersion: filter.kroxylicious.io/v1alpha1
+kind: Filter
+metadata:
+  name: myfilter
+spec:
+  type: io.kroxylicious.filter.encryption.RecordEncryption
+  config:
+    kms: Foo
+    kmsConfig: {}
+    selector: Bar
+    selectorConfig: {}
+```
+
+# Worked examples
+
+## On Cluster Traffic - plain downstream & upstream
+
+```yaml
+apiVersion: proxy.kroxylicious.io/v1
+kind: Proxy
+metadata:
+  name: myproxy
+spec: {}
+```
+
+```yaml
+apiVersion: proxy.kroxylicious.io/v1
+kind: ProxyIngress
+metadata:
+  name: myclusterip
+spec:
+  proxyRef:
+    name: myproxy
+  clusterIP:
+    protocol: TCP
+    port: 9082
+```
+
+```yaml
+apiVersion: proxy.kroxylicious.io/v1
+kind: VirtualCluster
+metadata:
+  name: mycluster
+spec:
+  proxyReference:
+    name: myproxy
+    kind: Proxy
+    group: proxy.kroxylicious.io
+
+  ingress:
+  - name: myclusterip
+
+  targetCluster:
+    resource:
+       bootstrap: upstream:9092
+       protocol: TCP
+       nodeIdRanges:
+      - name: mybrokers
+        range:
+          startInclusive: 0
+          endExclusive: 3
+  filterRef:
+  - group: filter.kroxylicious.io
+    kind: Filter
+    name: encryption 
+```
+
+What would operator create:
+* kroxylicious deployment, 1 replica
+* proxy relies on port based routing (needs new PortPerNode scheme that allows the Operator to control the ports precisely.  This is needed to prevent the potential for "crossed-lines" during reconcilations.)
+* 1 ClusterIP services (with n+1 ports) - operator needs to _deterministically_ assign a block of ports for this VC and create the port/target mapping in the Service accordingly.
+* Kafka Clients connect to serviceaddress:9082 
+
+## On Cluster Traffic - tls downstream & upstream
+
+```yaml
+apiVersion: proxy.kroxylicious.io/v1
+kind: Proxy
+metadata:
+  name: myproxy
+spec: {}
+```
+
+```yaml
+apiVersion: proxy.kroxylicious.io/v1
+kind: ProxyIngress
+metadata:
+  name: oncluster
+spec:
+  proxyRef:
+    name: myproxy
+
+  clusterIP:
+    protocol: TLS
+    port: 9082
+```
+
+```yaml
+apiVersion: proxy.kroxylicious.io/v1
+kind: VirtualCluster
+metadata:
+  name: mycluster
+spec:
+  proxyReference:
+    name: myproxy
+    kind: Proxy
+    group: proxy.kroxylicious.io
+
+  ingress:
+  - name: myclusterip
+    tls:
+      certificateRef:
+         ...
+
+  targetCluster:
+    resource:
+       bootstrap: upstream:9092
+       protocol: TLS
+       nodeIdRanges:
+      - name: mybrokers
+        range:
+          startInclusive: 0
+          endExclusive: 3
+
+  filterRef:
+  - group: filter.kroxylicious.io
+    kind: Filter
+    name: encryption 
+```
+
+What would operator create:
+* kroxylicious deployment, 1 replica
+* proxy relies on port based routing  (needs new PortPerNode scheme that allows the Operator to control the ports precisely.  This is needed to prevent the potential for "crossed-lines" during reconcilations.)
```

Comment:
This could use SNI based routing, but I figure since we've had to use port based routing for TCP, we may as well use port based routing for TLS too.